src/main/java/org/truffleruby/language/methods/LookupMethodNode.java

```diff
@@ -68,51 +72,83 @@ public InternalMethod lookup(VirtualFrame frame, Object self, String name) {
     protected InternalMethod lookupMethodCached(VirtualFrame frame, Object self, String name,
             @Cached("metaClass(self)") DynamicObject selfMetaClass,
             @Cached("name") String cachedName,
-            @Cached("doLookup(frame, self, name)") MethodLookupResult method) {
+            @Cached("doCachedLookup(frame, self, name)") MethodLookupResult method) {
         return method.getMethod();
     }
 
-    @Specialization
-    protected InternalMethod lookupMethodUncached(VirtualFrame frame, Object self, String name) {
-        return doLookup(frame, self, name).getMethod();
-    }
+    @Specialization(replaces = "lookupMethodCached")
+    protected InternalMethod lookupMethodUncached(VirtualFrame frame, Object self, String name,
+            @Cached("create()") MetaClassNode callerMetaClassNode,
+            @Cached("createBinaryProfile()") ConditionProfile noCallerMethodProfile,
+            @Cached("createBinaryProfile()") ConditionProfile isSendProfile,
+            @Cached("create()") BranchProfile foreignProfile,
+            @Cached("createBinaryProfile()") ConditionProfile foundProfile,
+            @Cached("createBinaryProfile()") ConditionProfile publicProfile) {
+
+        // Find the caller class
+        final DynamicObject callerClass;
+        if (ignoreVisibility || onlyLookupPublic) {
```

Comment:
Why not profile this branch as well?

Reply:
Those two are final fields of the node (the syntax highlighting unfortunately does not show it).